### DISCUSSION TITLE
Why keep nonce ?

### DISCUSSION THREAD

**Discussion by LionelLaffineur** [adminerevo/adminerevo#73] [Q&A]
I'm tired of that nonce which makes everything javascript related complex (too complex?).
Can someone give me a good reason not to remove it from the code ?

**Comment by gerryd** [comment]
The nonce works with the Content Security Policy (CSP), combined these will only allow loading verified sources. In other words, it prevents attacks where someone would (maliciously) inject external JavaScript code which has not been "signed".

See this sample `Content-Security-Policy` header returned by adminer in HTTP responses:

    Content-Security-Policy: script-src 'self' 'unsafe-inline' 'nonce-ZGJjODE5YzJlOTgwZTE0MjgyMjllMjhkNTJlOTdjMmM=' 'strict-dynamic'; connect-src 'self'; frame-src https://www.adminer.org/; object-src 'none'; base-uri 'none'; form-action 'self'

Note that we may want to consider removing `frame-src https://www.adminer.org/`, I don't know what it's being used for, maybe a version check or something?

**Reply by adrianbj** [reply]
@gerryd - just a note that I changed the `frame-rsc` to `self` when I redid the version checker: https://github.com/adminerevo/adminerevo/commit/07b3dc3f1d7af8f66b83c247cbe63926004c72ae

**Comment by LionelLaffineur** [comment]
OK, but if the nonce would "really" be useful for security, why isn't it used more ?
Is the added security worth the effort of the complexity and pain it brings ?
Because it is really painful !

**Reply by gerryd** [reply]
It is useful, but I agree that it may not always be convenient to use.

An environment like adminer is actually quite sensitive, it is being used to manage all the raw data which you are (usually) dearly trying to protect in your application's code. Having a leak there would have quite some impact. I think it makes sense to implement as many countermeasures as possible to prevent this from happening.

Now what might help alleviate your grievances, is maybe some sort of development mode where the nonce check is disabled?

**Reply by derStephan** [reply]
>OK, but if the nonce would "really" be useful for security, why isn't it used more ?

Because the normal developer thinks along the lines of your argument. Security makes the process more complicated. So this is oftentimes just omitted. 

I strongly vote against removing the nonce as it provides an effective, systemic, and fairly easy-to-use protection against XSS attacks. In the contrary, I vote for making the policy more strict and remove all exceptions. 

https://en.wikipedia.org/wiki/Content_Security_Policy

>Because it is really painful 

I don't get it. Where is your pain here? 

`<script <?php echo nonce(); ?> >...</script>`

That's all there is to do. 

>Is the added security worth the effort of the complexity and pain it brings ?

This is absolutely worth it because it forces us to develop secure software. 

Security was one of the main developing priorities for original adminer and I think, that should be kept for any fork that is supposed to be a replacement. If you lower that bar, people will have the false hope that you are as secure as the original adminer. 

>Now what might help alleviate your grievances, is maybe some sort of development mode where the nonce check is disabled?

This would mean that you develop something with lower security which will fail at once in production mode. You have to include security while developing. Always. Otherwise, people will start running everything in development mode - even in production.
